### DISCUSSION TITLE
clustering ideas

### DISCUSSION THREAD

**Discussion by priamai** [alexklibisz/elastiknn#364] [Q&A] [answered]
Hi there, I wonder whether would be possible to write an application to cluster documents as described in this old [article](https://www.sciencedirect.com/science/article/pii/S1877050916300199/pdf?md5=576a4d631094bd95808a4266834f3d17&pid=1-s2.0-S1877050916300199-main.pdf) .

It describes how k-means, k-medoid and k-means fast can be implemented via cosine similarity, jaccard coefficient and correlation coefficient. I believe would be trivial to also add euclidean similarity?

**Comment by alexklibisz** [comment, marked as answer]
You can't really do clustering with _only_ Elasticsearch/Elastiknn; you need some other system as well. 

Elasticsearch (and therefore also Elastiknn) is designed around a request-response model, dealing with a single document to index or a single query to execute. So you make a request and get the answer out of the cluster and back to the user as fast as possible. This means there's really no place to execute long-running iterative clustering techniques like K-means.

What you could do is this: run the clustering algorithm in some other system and use Elasticsearch/Elastiknn to store and retrieve them. So you basically start by running K-means for some representative subset of the document vectors, say in Python using scikit-learn. This gives you a smaller set of centroid vectors, one per cluster. You index all of the document vectors in Elasticsearch/Elastiknn and then use those centroid vectors to retrieve clusters of the document vectors (e.g., 100 nearest neighbors of a centroid vector is a cluster). You could also index the centroid vectors in another index and reference them in your queries against the document vectors.

If you have more follow-up questions, let's open a discussion: https://github.com/alexklibisz/elastiknn/discussions/new

**Comment by priamai** [comment]
Sure you are spot on, I did see some ES people attempting to produce an aggregator for k-means but never manage to complete, there are certain features missing in the ES engine for achieving that.

What I will try is the following: use Dask to load the ES documents (with the vectors) in parallel, run the k-means or k-medoids effcient implementations and then update the elements with a cluster id field.
This way at query time the user can retrieve the entire cluster. 
I was reading also about a new database pattern called "feature databases" where basically a database is designed to store ML features to be able to run ML jobs.
In this case means ES +EKNN fields are my feature database.

I will post some examples once is working.
Cheers!
